tests/utils.go

```diff
@@ -2096,11 +2097,17 @@ func AddEphemeralCdrom(vmi *v1.VirtualMachineInstance, name string, bus string,
 }
 
 func NewRandomFedora32VMIWithFedoraUser() *v1.VirtualMachineInstance {
-	vmi := NewRandomVMIWithEphemeralDiskAndUserdata(ContainerDiskFor(ContainerDiskFedora), `#!/bin/bash
-	    echo "fedora" |passwd fedora --stdin
-        echo `)
-	vmi.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory] = resource.MustParse("512M")
-	return vmi
+	userData := `#!/bin/bash
+echo "fedora" |passwd fedora --stdin
+echo `
+	return libvmi.NewDefault(
```

Comment:
1. Having a method called `NewDefault` that gets parameters is very confusing (especially when two of the params passed here have "default" in their name). I would expect `NewDefault` to create a default vmi a not getting any param.
Something like this - `NewDefault().WithResourceMemory("512M").With(Rng())...CreateInstance()
Or at least -
baseVmi := libvmi.NewDefault()
libvmi.CreateVmi(baseVmi, libvmi.WithResourceMemory("512M"), libvmi.WithRng() ...)
2. I understand the logic behind this PR and I think it can be very useful that our vmi factory methods will look like the PR suggests, with the `with` building blocks. And if a special vmi is needed it can be easily constructed with this building blocks.
However NewRandomVMIWithEphemeralDiskAndUserdata is not special. It has 101 use cases in the tests.
All those use cases don't care about the Rng, the name, the namespace and the b64Encoding of the userData.
I would expect the code here to look something like this -
baseVmi = NewRandomVMIWithEphemeralDiskAndUserdata(ContainerDiskFor(ContainerDiskFedora), `#!/bin/bash
	    echo "fedora" |passwd fedora --stdin
        echo `)
return libvmi.CreateVmi(baseVmi, libvmi.WithResourceMemory("512M"))

Reply:
>     1. Having a method called `NewDefault` that gets parameters is very confusing (especially when two of the params passed here have "default" in their name). I would expect `NewDefault` to create a default vmi and not getting any param.
>        Something like this - `NewDefault().WithResourceMemory("512M").With(Rng())...CreateInstance()
>        Or at least -
>        baseVmi := libvmi.NewDefault()
>        libvmi.CreateVmi(baseVmi, libvmi.WithResourceMemory("512M"), libvmi.WithRng() ...)

The defaults represents the current code, I have not changed that logic.
They come from the common of all VMI creations. The fact that you can recognize this here is the whole point of this change.
I think it will be easier to convert other usages this way and only then take the `NewDefault` out of the picture.
Refactoring cannot be done in one shot.

Not accepting any parameters and then building it with some other builder functions is exactly what I wanted to avoid by this change. This is how things are built now and it looks terrible and very hard to read... each layer adds its own stuff and to understand what we have one needs to go all the way down the stack.
I see no problem with creating the default VMI and adding to it all options one needs. This explicitly causes the caller to build the VMI in one place for the reader to see.

> 
>     2. I understand the logic behind this PR and I think it can be very useful that our vmi factory methods will look like the PR suggests, with the `with` building blocks. And if a special vmi is needed it can be easily constructed with this building blocks.
>        However NewRandomVMIWithEphemeralDiskAndUserdata is not special. It has 101 use cases in the tests.
>        All those use cases, currently, don't need care about the Rng, the name, the namespace and the b64Encoding of the userData.

The users have no idea what they get with that factory function, so they explicitly overwrite options because they do not trust what is configured down the stack.
If one wants to use such a function, they can still do it and explicitly build the VMI with all the options, including the RNG, hiding it from the user. But if the user wants to see what it means, he just needs to look one layer down and not 10.

> 
> 
> I would expect the code here to look something like this -
> 
> ```
> baseVmi = NewRandomVMIWithEphemeralDiskAndUserdata(ContainerDiskFor(ContainerDiskFedora), `#!/bin/bash
> 	    echo "fedora" |passwd fedora --stdin
>         echo `)
> return libvmi.CreateVmi(baseVmi, libvmi.WithResourceMemory("512M"))
> ```

This will keep the current state and make this change irrelevant. Maybe a bit nicer, but not changing how I expect developers to use it.
I expect users to build specific factories that use the `libvirt.New` with the `With*` options.
I do not want them to edit the VMI after through a different API.
The factories, like the `NewDefault` can collect options on their way, but I think it will be much nicer to collect a collection of options and name them so factories or tests directly can use them.